### ISSUE TITLE
It is an error to have a setter with the same name as the enclosing class

### ISSUE THREAD

**Issue by eernstg** [dart-lang/sdk#34225]
A class name can clash with the basename of an instance or static setter (cf. 9085009f2a2fdfec6172488486e47997e062fe5a):

```dart
class C {
  static set C(v) {} //# 01: compile-time error
  set C(v) {} //# 02: compile-time error
}
```
`dart` (version 2.1.0-dev.0.0) reports an error for 01 (but not the correct one, it says `Constructors can't be static`) and accepts 02. `dartanalyzer` (version 2.1.0-dev.0.0) does not report an error for 01 nor 02.

Note that this is a near-revert of 402e69b53d59083d34d7538126837c4381aa0c4a which was used to close #30834: This error used to be emitted by the analyzer until that commit.

This means that a fix to this issue introduces an error which is a breaking change, but it was also an error according to the analyzer until Feb 6, 2018, so the breakage is not likely to be large.

**Comment by matanlurey** [comment]
Is this a duplicate of (or overlap with) https://github.com/dart-lang/sdk/issues/32968 @eernstg?

**Comment by eernstg** [comment]
No, #32968 is concerned with a conflict where a superclass declares a static member and a subclass declares a member (instance of static) with the same basename. This issue is concerned with a conflict between the name of a class and the basename of a setter.

**Comment by matanlurey** [comment]
Ah ok, _carry on_.

**Comment by bwilkerson** [comment]
Created https://github.com/dart-lang/sdk/issues/34289 to track the analyzer work.

**Comment by peter-ahe-google** [comment]
@eernstg FYI: This is a breaking change from Dart 1. The name of the setter above is "C=" so it didn't use to be an error.

**Comment by peter-ahe-google** [comment]
I'm not worried about the breakage. I'm just pointing out that your statement that this was an error until Feb 6, 2018 isn't correct.

**Comment by eernstg** [comment]
Ah, right, the Feb 2018 change in 402e69b53d59083d34d7538126837c4381aa0c4a fixed a kernel code generation problem, it did not eliminate an incorrect error (which is now a correct error).

But the other justifications for the change should still be valid.
